ruby/ruby-agreement.html

```diff
@@ -0,0 +1,45 @@
+<!doctype html>
+<html lang="en">
+<meta charset="utf-8">
+<meta name="viewport" content="width=device-width">
+<title>Agreement on HTML Ruby Markup</title>
+<style>
+body {
+	max-width: 50em;
+	margin: auto;
+	font-family: sans-serif;
+}
+</style>
+
+<h1>Agreement on HTML Ruby Markup</h1>
+
+<p>
+The W3C will specify the extended HTML Ruby markup
+described in <a href="https://github.com/whatwg/html/pull/6478">PR#6478</a>
+as a <a href="https://www.w3.org/Consortium/Process/#rec-track">REC track</a> document.
+The W3C and WHATWG believe this is within the bounds of our <a href="https://www.w3.org/2019/04/WHATWG-W3C-MOU">MoU</a>,
+and while it is not the ideal outcome
+(having Ruby documented in two places),
+we understand that it serves the goals for the W3C’s constituencies and existing implementations
+to have an easily referencable document
+as well as honoring the strict <a href="https://whatwg.org/working-mode#additions">requirements of the WHATWG</a>
+that the living standard not contain new additions
+that don’t have multi-browser-engine implementer commitments.
+
+<p>
+The <abbr title="Status Of This Document">SOTD</abbr> section of the W3C Ruby Extension spec
+will state our joint intention
+that the extended ruby markup definitions be fully integrated into the <a href="https://html.spec.whatwg.org/multipage/">WHATWG HTML specification</a>
+once they <a href="https://whatwg.org/faq#adding-new-features">meet WHATWG’s inclusion criteria</a>
+under section <a href="https://www.w3.org/2019/04/WHATWG-W3C-MOU#forking">10.2 of the 2019 MOU</a>.
+
+<p>
+W3C editors will offer pull requests
+to keep the HTML spec in sync,
+both technically and editorially,
+for the subset of features defined in both specs
+(goal is to reduce the deltas
+such that the only differences will be
+the addition of the two new elements <code>&lt;rb></code> and <code>&lt;rtc></code>);
+WHATWG HTML editors will work with the W3C Ruby Extension spec authors
+to merge those pull requests in support of that goal.
```

Comment:
I think that's reasonable... and is likely otherwise less work for W3C editors. How about changing the paragraph to something like:
> W3C editors may offer pull requests that correct existing Ruby algorithms for which browser implementations are not aligned, along with accompanying web platform tests per WHATWG working mode process. It is not expected that the W3C document and WHATWG spec content be kept in sync prior to being fully integrated under the conditions noted previously. WHATWG HTML editors will evaluate and optionally merge such PRs on their individual merit.

?

Reply:
This would be
* More complicated editing work for the W3C editors, because they would need to keep two things normatively synchronized which are not editorially synchronized.
* Many hours of additional effort to explain the nuances of each minor adjustment and clarification to the prose in order to justify each individual change as an improvement, instead of simply defending the correctness of a complete rewrite and the editorialness or correctness of any subsequent fixes.
* Confusing to readers and implementers, because it would not be obvious what the normative differences are between the two specs.
* A disservice to WHATWG readers, because improvements to examples and to the prose's readability will not be available to them. (And improvements are needed, this part of the spec is not well-written and its examples are not particularly well-explained, nor all necessarily good practice.)

So I don't think that's a positive change to our agreement.

Florian and I are willing to submit PRs for full synchronization changes, as agreed with the Steering Group. If WHATWG wants to backtrack on our agreement and not merge them, well, we believe it's a poor decision but can probably live with it: having the rest of the agreement still moves us forward. But in that case we'll leave the work of maintaining this section of the WHATWG spec to the WHATWG editors.

However, we would rather see this section of the WHATWG spec be something people want to reference, rather than being a poorly-explained shadow of the W3C spec.

Reply:
It seems like there are larger disagreements here on the direction that were not discussed. I don't think this is quite at the "agreement" stage yet, in that case.

Reply:
In particular the unwillingness to do extra complicated editing work and explanation is disappointing, and not the spirit I thought this agreement was in. Just lobbing changes over and expecting them to be accepted without evaluation, explanation, or collaboration is not a reasonable way to interface with the WHATWG.

Reply:
> expecting them to be accepted without evaluation, explanation, or collaboration

We do not expect changes to be accepted without evaluation, explanation, or collaboration. We expect you to evaluate them, for us to explain anything you don't understand, and for us to process your feedback and suggestions for improvements.

But I, at least, am not willing to maintain a second, editorially incompatible version of a spec because you are unwilling to accept editorial improvements that, to you, do not weigh more heavily than your “don't make any unnecessary changes” bias. That's a reasonable position to take for the rest of your spec, but here we would like to replace it with a “don't have any unnecessary differences” bias.

Reply:
Alright. I understand the disagreement.

Given the overall lack of collaboration, maybe it's best if there's no formal agreement here, and the W3C work on Ruby is just treated like any other new-element incubation. Like, we don't need a formal agreement with the W3C for the portals spec. I think this ruby work should be treated on the same level, personally.

Reply:
I still believe it is valuable to record that we've agreed that having the two docs is OK (should anyone come asking), and to have the note about ultimately desiring to merge the Ruby extension back in--even if it is largely understood that this is the general process for other incubation work (e.g., portals, as noted).

May I propose then, that we all agree to simply drop this third paragraph about PR contributions? Given the above discussion, it sounds like it would be more work for the W3C editors after all, and is not an essential part of the proposed Ruby extension spec work.

Reply:
Maybe we could make it a more general statement about incubations being OK, instead of creating this document just for one particular case?

Reply:
Perhaps, though I suspect we'd run into a thorny patch when trying to describe exactly what an incubation means :-)

I think this case is somewhat unique in that it has a lot of history between the two orgs and the various participants, is already specified in one way in the HTML standard, and the intended "incubation" will be a Rec-track journey in the W3C's HTML WG. Rather than shift the focus of the statement, my desire is to land this statement as we mutually understand it--I and think we've arrived at something more-or-less agreeable. Should something else come up with another incubation effort then we cross that road when we come to it. As you've noted, it seems like other incubations are proceeding without concern--making this into statement about incubations may be superfluous.

Reply:
Well, I have a bit more concern about this incubation than others, given the unwillingness to contribute back unless changes are accepted wholesale. So sanctioning it with a special agreement, while not giving the same consideration to more cooperative incubations, seems not-great to me.

Reply:
@fantasai and I spoke offline. As usual, such conversations are helpful for turning down the temperature, and I want to thank her for reaching out to initiate it.

We discussed how there's a tradeoff between how much time @fantasai and @frivoal are willing to spend working on both their document and on pull requests, versus the usual expectations when submitting pull requests to WHATWG specs. Our tentative takeaway was that the best path to success would probably involve submitting a series of pull requests to the HTML Standard, in order from most-likely-to-be-accepted-with-minimal-fuss to trickiest:

- Fix normative errors in the existing spec, with accompanying web platform tests.
- Update or expand the examples. (This would require a brief explanation of what was wrong with the existing ones, but I wouldn't expect to get into an extended back-and-forth.)
- Add a non-normative summary of the algorithms. (@frivoal's PR removes the algorithms entirely, replacing them with a couple paragraphs, which I am not excited about. But summarizing the intent of the algorithms non-normatively would be a very positive change. This would be a good example of collaboratively adapting to the conventions of the destination document, which prefers algorithms more than CSSWG documents seem to.)
- Update introductory definitions and examples, e.g., @frivoal's PR actually states what the `ruby` element represents which is a good change from the existing HTML Standard which does not. (This will require explaining the changes to make sure we understand why the new ones are better. E.g. it is not clear to me why @frivoal changed `rp` outside `ruby` from representing its contents to representing nothing, or why in his PR `rt` represents something instead of contributing to what `ruby` represents. On the other hand, some changes, such as just adding links to more informative documents like [QA-RUBY], are clearly good.)

There may be other notable chunks but it's hard for me to understand currently how much of the changes are tied up in the rb/rtc model vs. which fit with the existing model.

The result of any of these might be disagreement, especially toward the bottom of the list. At that point it's up to the contributors whether or not they want to spend the time working through the disagreement, or not. If it's not worth the extra hours to get to agreement, and instead is only worth submitting PRs which are quickly approved and merged, then sticking toward the top of the list is probably the best use of time.

We also briefly discussed how we could avoid some of the desire for editorial synchronization by, instead of duplicating text across both locations, removing the text from the extension spec and instead linking to paragraphs, examples, definitions, etc. in HTML. That would fit better with how most other extension incubations work. We didn't explore that in detail though so I'm not sure there was agreement.

---

Regarding the tactical question of the Agreement: I am still unsure that an official agreement about this work is a good idea. Even with collaboration along the lines above, I am uneasy about singling out this particular incubation work compared to others. Is this to be read as saying the procedure we're following here is unusual? I hope not. I would prefer a more general statement about how incubation extension to WHATWG specs are expected to work.

But, I'll note that with the last paragraph removed, the agreement only states things the W3C will do. So maybe one path would be for it to become a W3C-initiated statement-of-intentions, instead of something the WHATWG signs off on. That is, it's up to individual W3C-hosted incubations whether they want to publish such statements-of-intention, and Ruby has chosen to do so.

Reply:
I appreciate @domenic taking the time to chat offline, I think it was very helpful in understanding what we can do to coordinate going forward. Teasing apart individual decisions the wording of each sentence into a separate diff and writing a rhetorical analysis essay explaining the editorial justification for each modification as I had to do [for Scroll Snap](https://drafts.csswg.org/css-scroll-snap/diff-notes#change-reword) is far beyond what I think is reasonable to require, but it seems (?) that we can split up the PR into larger chunks than that, and if there are questions about specific things like the representation of `rp`, it's no problem to answer such questions.

@domenic and I both think one of the trickiest such chunks is likely to be the rewrite of the pairing algorithm, since we changed entirely how, editorially speaking, the spec approaches the problem, and it may be that this section remains as-is for now in the WHATWG spec. But if he's willing to review it, it might be worth a try to sync this part also. If nothing else, such review would help verify the correctness of both versions (or find whatever problems may exist).

Wrt the suggestion of linking to paragraphs/examples/definitions in HTML instead of duplicating them, I think it would make it difficult to read and understand the extension spec if those pieces weren't incorporated into the text, since they're integral to understanding what's happening--and ruby is already something most readers are unfamiliar with, so optimizing for understandability is imho worth duplicating text. (It may be that for other extension specs there's cleaner ways of integration, but here it would involve some very awkward monkey-patching in multiple places.)

Regarding having the agreement as drafted here, I agree with @travisleithead that this is worth documenting. The differences in the ruby markup model have been a singular point of contention between WHATWG and W3C for many years, and unlike other incubations, the intent at W3C is to take it through the standards track to REC, so, not just an incubation from a W3C point of view and something that will need to be chartered into the HTMLWG. Ensuring that the WHATWG approves of the HTMLWG doing that (and documenting this publicly) is the whole reason we're having this conversation, in order to head off any potential friction.

Wrt the third paragraph, I'd like to hear from the SG what they think, as it was one of our goals to do as much as we could to avoid any implication of “forking” the HTML spec rather than extending it. The main point of disagreement is, I think, whether synchronization with W3C text or maintaining the stability of WHATWG's existing text is more important when evaluating editorial changes.

But in any case we should clarify that “WHATWG HTML editors will work with the W3C Ruby Extension spec authors” means that the WHATWG editors are expected to review the PRs on their merits, and the Extension spec authors are expected to address any feedback to ensure that the PRs conform to WHATWG's expectations of technical correctness and editorial clarity before merging.

Reply:
I think the most important thing here is that we document in public that the WHATWG acknowledges that the Ruby extension spec exists, that there has been "good faith negotiation" and that this not considered a fork outside the bounds of the MoU. The first paragraph achieves that.

When it comes to what changes will be made to the HTML spec, would it be acceptable to state that there will be good faith collaboration, while following our usual working mode? Suggestion:

> W3C editors may offer pull requests to keep the HTML spec in sync, both technically and editorially, for the subset of features defined in both specs. WHATWG HTML editors will review these contributions in good faith, following the WHATWG [working mode](https://whatwg.org/working-mode).

Unstated would be that if collaboration breaks down, appealing to the SG is still possible.

I recognize it would be nice to agree up front on what the HTML spec's Ruby section is going to be like, but it sounds like that will depend on the amount of time and effort that the W3C+WHATWG editors thinks it's worth when getting into the details.

Reply:
Thanks, I think this is a good suggestion that works a reasonable balance between what everybody has expressed so far.

Reply:
I have just pushed new commits. The first is simply drops the original third paragraph and replaces it with the one @foolip just suggested.

After that, as long as we're making some adjustments, I have included a few more minor wording tweaks for the sake of clarity. I believe they do make no difference to the what we are agreeing to, merely make the text easier to understand and help steer clear of undesired implications. For ease of review, each change is a separate commit, with the justification in the commit message.

Reply:
Let me know if this is now acceptable and ready to merge